### PR TITLE
[!!!][BUGFIX] Change Glossary functionality

### LINKED ISSUE #99: Glossary entries are not used when content is translated - because of bug in DataHandlerHook, glossary entries are not synced
**Update**: The problem could be tracked down, see comments below, see PR #100.

-------

In general, it would be nice if there were a status check or one would get some feedback if the glossary is being used.

Is there something I can do to troubleshoot this or is something missing? Is further configuration necessary?

Entries "DeepL Glossaries Sync" are not automatically created (our contact in another university gave this tip, that this should happen).

In the (logging framework) logs there is no error during the timeframe where the entries were created or translations took place. 

## Reproduce 

Followed information in README: https://github.com/web-vision/wv_deepltranslate#deepl-glossary 

1. Created page A for translation with default language = German and translation (sys_language_uid=1) English
2. Created subpage B for glossary entries, selected in page properties tab "Behaviour" | "Use as Container" | "Contains Plugin": **"Deepl Glossar"**
3. Switched to list module and added glossary entries:
    - de: Fakultät, en: "school"
    - de: Semesterferien, en "lecture-free period"

image

 4. On page A, translated existing text to English using "**Translate (deepl)**" (not "Translate (deepl) (autodeted)")

image

> Für die **Fakultät** gibt es keine Studiengänge.
> Momentan haben wir **Semesterferien**.

The result is:
> There are no courses of study for the **faculty**.
> At the moment we have **semester break**.

I would expect school instead of faculty and "lecture-free period" instead of "semester break"

## version

* current master wv_deepltranslate
* PHP 7.4
* TYPO3 11.5.21

or

* current master wv_deepltranslate
* PHP 8.1
* TYPO3 11.5.22

because of #98 the code was patched:

```
git diff
diff --git a/Classes/Hooks/DataHandlerHook.php b/Classes/Hooks/DataHandlerHook.php
index 64bb139..880fea0 100644
--- a/Classes/Hooks/DataHandlerHook.php
+++ b/Classes/Hooks/DataHandlerHook.php
@@ -95,7 +95,7 @@ class DataHandlerHook implements LoggerAwareInterface
         if ($status !== 'update') {
             return;
         }
-        if ($fieldArray['l10n_parent'] === 0) {
+         $l10n_parent = $fieldArray['l10n_parent'] ?? false;
+        if ($l10n_parent === 0) {
             return;
         }

```

## Settings

I can't show you my API key, but the rest is:

image

image

image

site configuration:

```
languages:
  -
    title: Deutsch
    enabled: true
    languageId: 0
    base: /
    typo3Language: de
    locale: de_DE.UTF-8
    iso-639-1: de
    navigationTitle: Deutsch
    hreflang: de-DE
    direction: ltr
    flag: de
    websiteTitle: ''
  -
    title: English
    enabled: true
    languageId: 1
    base: /en/
    typo3Language: default
    locale: en_UK.UTF8
    iso-639-1: en
    navigationTitle: English
    hreflang: en-GB
    direction: ltr
    fallbackType: strict
    fallbacks: ''
    flag: gb
    websiteTitle: 'University of Oldenburg'


```

### LINKED ISSUE #114: Feature request: Make it possible to configure the glossary name
Currently, the glossary name is constructed by

```
$glossaryNamePrefix = 'DeepL';
$glossaryName = $glossaryNamePrefix . '-' . strtoupper($sourceLang) . '-' . strtoupper($targetLang);
```

Effectively, you cannot control the glossary name. If you test the glossary on a test system, you might unintentionally overwrite the glossary on the production system.

### LINKED ISSUE #117: Synchronising of glossary does not work, glossary is not used (and it is not documented how it should work)
I think this might have worked before, because otherwise #99 would not have occurred. Also, the list command shows several glossaries (which is also not correct because there should not be several), the latest of them was January 13, 2023.). For further tests, I deleted the already existing glossaries on DeepL so that can't be the reason this is not working anymore.

I have debugged the code while creating glossary entries and they are never synchronized (entries in tx_wvdeepltranslate_domain_model_glossariessync are not created and glossary is not created on DeepL) and it seems obvious it cannot work, at least not in v11.

I set a breakpoint in DataHandlerHook::prepareLangagues, DataHandlerHook::processDatamap_afterDatabaseOperations and GlossarySyncController::update, DeeplGlossaryService::createGlossary

In order to synchronized glossaries, a hook is used: DataHandlerHook::processDatamap_afterDatabaseOperations.

If the following is true, it returns and does nothing

* `if ($status !== 'update') {`
* `if (!isset($fieldArray['l10n_parent']) || $fieldArray['l10n_parent'] === 0) {`

When debugging this, it always returns because `l10n_parent` is not set.

When TYPO3 calls this hook, it does not pass the full field array, it passes only the changed values, see core DataHandler, before calling processDatamap_afterDatabaseOperations:

 $fieldArray = $this->compareFieldArrayWithCurrentAndUnset($table, $id, $fieldArray);

Unfortunately, the hook is not well documented.

Also, I do not think it is a good idea to sync 

## Versions

* TYPO3 11.5.22
* wv_deepltranslate: latest master with patch for #112

## current list of glossaries

The (undocumented) command shows:

```
php vendor/bin/typo3 wv_deepltranslate:glossary_list
============
Read more here: https://www.deepl.com/docs-api/managing-glossaries/listing-glossaries/
============
+--------------------------------------+-------------+-------+-------------+-------------+-----------------------------+-------------+
| glossary_id                          | name        | ready | source_lang | target_lang | creation_time               | entry_count |
+--------------------------------------+-------------+-------+-------------+-------------+-----------------------------+-------------+
| fd9d3e36-b49b-46ab-bf4e-6e496ec86f8b | DeepL-DE-EN | 1     | de          | en          | 2023-01-13T14:23:05.777668Z | 1           |
| 7f6012e5-f92d-42c8-ad47-16bc31ae0fe3 | DeepL-DE-EN | 1     | de          | en          | 2023-01-13T15:16:51.6142Z   | 1           |
| 82f145f2-3b0c-405f-a4c7-a798f44ac080 | DeepL-DE-EN | 1     | de          | en          | 2022-08-09T07:00:04.146722Z | 1           |
| 4cd291f9-059c-4d11-a0b2-a0f6befd9f72 | DeepL-DE-EN | 1     | de          | en          | 2022-08-09T07:00:21.947595Z | 1           |
| b3b071ae-c715-4503-ae8c-fd14f124e1dc | DeepL-DE-EN | 1     | de          | en          | 2022-08-09T07:11:25.170876Z | 1           |
| 1064e7a0-d0a0-4635-aaa0-0d21ccc423ad | DeepL-DE-EN | 1     | de          | en          | 2022-08-09T07:11:41.074742Z | 1           |
| ae9c3db9-c53c-4b80-a48c-52d0e16e9a2c | DeepL-DE-EN | 1     | de          | en          | 2022-08-15T13:08:00.366722Z | 1           |
| 358faacf-9c69-466a-aa73-8163c0d00ca2 | DeepL-DE-EN | 1     | de          | en          | 2022-08-15T13:08:10.57717Z  | 2           |
| 21721f2c-2812-46cb-8555-29deb4c9ea0c | DeepL-DE-EN | 1     | de          | en          | 2022-08-15T13:08:58.83386Z  | 2           |
+--------------------------------------+-------------+-------+-------------+-------------+-----------------------------+-------------+

```

Probably during previous tests, these were created.

After deleting them, no new glossaries are uploaded to DeepL, even after creating new glossary entries and translating something. 

None of the breakpoints mentioned above are reached, except for DataHandlerHook::processDatamap_afterDatabaseOperations where the function is left right away because l10n_parent is not set. Also, no entrie in tx_wvdeepltranslate_domain_model_glossariessync are created.

```
php vendor/bin/typo3 wv_deepltranslate:glossary_cleanup
php vendor/bin/typo3 wv_deepltranslate:glossary_list
============
No Glossaries found.
Read more here: https://www.deepl.com/docs-api/managing-glossaries/listing-glossaries/
============

```

```
select count(*) from  tx_wvdeepltranslate_domain_model_glossariessync;
+----------+
| count(*) |
+----------+
|        0 |
+----------+
```

### PR DESCRIPTION
**Pull request by calien666**
After reviewing the code, Glossary functionality was a crap.

Complete rework and restructuring of usage for glossaries.

New tables defined, which can handle glossaries correctly. One Glossary now contains n entries. Per glossary the target and source will be set, with respecting the possible glossary features from DeepL.

Glossary has information about last sync to DeepL and checks, whether changes are inside the Glossary or not. 

fixes #117
fixes #99 
fixes #114 

Screenshot 2023-02-21 at 02 45 01
Glossary overview with Sync button on top

Screenshot 2023-02-21 at 02 47 42
View of glossary entries, entry opened

Screenshot 2023-02-21 at 02 46 12
View of glossary entries, entries closed

Screenshot 2023-02-21 at 02 45 48
Synchronisation information

Screenshot 2023-02-21 at 02 45 27
Glossary not synced after change.

### CODE REVIEW

File: Classes/Domain/Repository/GlossaryRepository.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    /**
+     * @param int $uid
+     * @return array{name: string, id: string, source_lang: string, target_lang: string, entries: array}
+     */
+    public function getGlossaryInformationForSync(int $uid): array
+    {
+        $glossaryInformation = [
+            'entries' => [],
+        ];
+        $db = GeneralUtility::makeInstance(ConnectionPool::class)
```
Comment (helsner) [inline Classes/Domain/Repository/GlossaryRepository.php:23]: injecting ConnectionPool or creating a property seems useful to me, as it is used multiple timees

File: Classes/TCA/LanguageSelectorItemsProc.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+use TYPO3\CMS\Backend\Utility\BackendUtility;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Extbase\Utility\LocalizationUtility;
+use WebVision\WvDeepltranslate\Service\DeeplGlossaryService;
+
+class LanguageSelectorItemsProc
+{
+    public function getFieldsForTarget(array &$configuration): void
+    {
+        $glossaryService = GeneralUtility::makeInstance(DeeplGlossaryService::class);
```
Comment (helsner) [inline Classes/TCA/LanguageSelectorItemsProc.php:16]: maybe use DI as well?

File: Classes/Domain/Repository/GlossaryRepository.php (hunk 1 above)
Comment (sbuerk) [reply Classes/Domain/Repository/GlossaryRepository.php:23]: as long as v9 is involved, DI is not really possible - needs fallback to GU::mI() then .. but generally a valid point.

File: Classes/TCA/LanguageSelectorItemsProc.php (hunk 2 above)
Comment (sbuerk) [reply Classes/TCA/LanguageSelectorItemsProc.php:16]: as long as v9 is involved, DI is not really possible - needs fallback to GU::mI() then .. but generally a valid point.

**Review by helsner — APPROVED** [review]
i only checked code, so no testing review, but from what I understand it looks good and makes sense.
But please stick to others, more experienced reviewers 😅

**Comment by artus70** [comment]
Looks nice! I'd like to have a natural sorting - either by source or by target. Should be helpful when having hundrets of entries. Otherwise it'd be difficult to find certain entries again.

**Comment by calien666** [comment]
> Looks nice! I'd like to have a natural sorting - either by source or by target. Should be helpful when having hundreds of entries. Otherwise it'd be difficult to find certain entries again.

Valid point. I'll add this. I guess, working inside a glossary should have sorting default by source, ascending?

**Comment by ahobitt** [comment]
> > Looks nice! I'd like to have a natural sorting - either by source or by target. Should be helpful when having hundreds of entries. Otherwise it'd be difficult to find certain entries again.
> 
> Valid point. I'll add this. I guess, working inside a glossary should have sorting default by source, ascending?

The default sorting should be ascending. That would be great.

**Review by sypets — COMMENTED** [review]
With this branch, I get error message on executing the "DB compare":

With latest TYPO3 v11 release (11.5.24)

```
$ composer require web-vision/wv_deepltranslate:dev-bugfix/glossary-sync
$ php vendor/bin/typo3cms database:updateschema
```
```
$ php vendor/bin/typo3cms database:updateschema
No schema updates were performed for update types:
"field.add", "field.change", "table.add", "table.change"

The following errors occurred:
+------------+--------------------------------------------------------+-----------------------------+
| Type       | SQL Statement                                          | Message                     |
+------------+--------------------------------------------------------+-----------------------------+
| Add fields | ALTER TABLE `tx_wvdeepltranslate_glossaryentry` ADD    | Incorrect table definition; |
|            | `uid` INT UNSIGNED AUTO_INCREMENT NOT NULL             | there can be only one auto  |
|            |                                                        | column and it must be       |
|            |                                                        | defined as a key            |
|            |                                                        |                             |
|            | ALTER TABLE `tx_wvdeepltranslate_glossaryentry` ADD    | Key column 'uid' doesn't    |
|            | PRIMARY KEY (uid)                                      | exist in table              |
|            |                                                        |                             |
|            | ALTER TABLE `tx_wvdeepltranslate_glossary` ADD `uid`   | Incorrect table definition; |
|            | INT UNSIGNED AUTO_INCREMENT NOT NULL                   | there can be only one auto  |
|            |                                                        | column and it must be       |
|            |                                                        | defined as a key            |
|            |                                                        |                             |
|            | ALTER TABLE `tx_wvdeepltranslate_glossary` ADD PRIMARY | Key column 'uid' doesn't    |
|            | KEY (uid)                                              | exist in table              |
|            |                                                        |                             |
+------------+--------------------------------------------------------+-----------------------------+
Script @php vendor/bin/typo3cms database:updateschema handling the typo3-post-deploy event returned with error code 1

```

Field uid was not created.

```
describe tx_wvdeepltranslate_glossaryentry;
+-----------+-------------------+------+-----+---------+-------+
| Field     | Type              | Null | Key | Default | Extra |
+-----------+-------------------+------+-----+---------+-------+
| source    | varchar(255)      | YES  |     | NULL    |       |
| target    | varchar(255)      | YES  |     | NULL    |       |
| glossary  | int               | YES  |     | NULL    |       |
| pid       | int unsigned      | NO   | MUL | 0       |       |
| tstamp    | int unsigned      | NO   |     | 0       |       |
| crdate    | int unsigned      | NO   |     | 0       |       |
| cruser_id | int unsigned      | NO   |     | 0       |       |
| deleted   | smallint unsigned | NO   |     | 0       |       |
+-----------+-------------------+------+-----+---------+-------+
8 rows in set (0.00 sec)

```

**Comment by calien666** [comment]
@sypets I'll take a look.

Did you checkout an older version of this branch for testing?

Only thing this could have happened is, that in an older version I forgot to clearly define TCA right, so TYPO3 couldn't set up database correct.

I tried with latest branch version on TYPO3 v11 in a system, where tables wheren't present and tables where created correct.

**Comment by calien666** [comment]
> > > Looks nice! I'd like to have a natural sorting - either by source or by target. Should be helpful when having hundreds of entries. Otherwise it'd be difficult to find certain entries again.
> > 
> > 
> > Valid point. I'll add this. I guess, working inside a glossary should have sorting default by source, ascending?
> 
> The default sorting should be ascending. That would be great.

Done with latest commits.

**Comment by sypets** [comment]
I dropped tables `tx_wvdeepltranslate_glossary` and `tx_wvdeepltranslate_glossaryentry` and executed the db command again with latest version and was ok.

**Review by sypets — CHANGES_REQUESTED** [review]
After installing this branch, I get exceptions in the page layout (just loading it, not related to translating). 

Reproduced with:
- clone of main site (with other extension) with TYPO3 v11.5.24 and PHP 8.1
- TYPO3 core development site based on latest 11.5 branch with no other extension (except styleguide) and PHP 8.1

I deinstalled wv_deepltranslate first, performed DB updates, flushed system caches and then installed the branch.

This breaks backend functionality (without even using the glossary or translating anything).

(1)
```
#1476107295 TYPO3\CMS\Core\Error\Exception
PHP Warning: Array to string conversion in /var/www/t3coredev11/vendor/typo3fluid/fluid/src/Core/Parser/SyntaxTree/AbstractNode.php line 73

```
I debugged and looked at the passed $value in castToString:

```
value
0 => 'Neue Übersetzung dieser Seite erstellen'
/typo3/record/commit?token=f5c48333dd5e0d1ab103e47ac67fabcb7369d7d8&redirect=%2Ftypo3%2Frecord%2Fedit%3Ftoken%3D9a2f0a1c119f048a0f994cf550489f24d9bfd4e0%26justLocalized%3Dpages%253A3%253A2%26returnUrl%3D%252Ftypo3%252Fmodule%252Fweb%252Flayout%253Ftoken%253D9199032a6eef2ccb404d82e3b4f719144fe5e18c%2526id%253D3%2526&cmd%5Bpages%5D%5B3%5D%5Blocalize%5D=2&cmd%5Blocalization%5D%5Bcustom%5D%5Bmode%5D=deepl 
=> 'Nederlands'
...
more entries for each language
```

**Comment by calien666** [comment]
@sypets Will take a look. This shouldn't happen

**Comment by sypets** [comment]
Update, the exception can be reproduced with latest master, created issue #150

**Review by NarkNiro — APPROVED** [review]
👍🏻

**Pull request merged**